### PR TITLE
Client Restore State Fixes

### PR DESCRIPTION
**Pull request by dadgar**
This PR updates how we restore state such that if a handle is not able to be opened we do not fail launching the client. Instead we now attempt to open the handle and if it fails, launch a new task_runner. This PR also re-enables the RestoreState and Open tests.

Fixes: #135, #176

### CODE REVIEW

File: client/allocdir/alloc_dir.go [hunk 1]
```diff
@@ -126,18 +126,24 @@ func (d *AllocDir) Embed(task string, dirs map[string]string) error {
...
 		}
 
-		// Enumerate the files in source.
-		entries, err := ioutil.ReadDir(source)
-		if err != nil {
-			return fmt.Errorf("Couldn't read directory %v: %v", source, err)
+		// Check if destination directory exists. This can happen if restarting
+		// a failed task.
+		destDir := filepath.Join(taskdir, dest)
+		if _, err := os.Stat(destDir); err == nil {
```
Comment (cbednarski) [inline client/allocdir/alloc_dir.go]: What happens if nomad failed halfway through copying the files into place? For the sake of idempotency it might be best to run through the rest of the procedure here.

Comment (dadgar) [inline client/allocdir/alloc_dir.go]: Yeah I can push this check down to the individual entries. Just need to avoid calling linkOrCopy on an existing link/copy as that will fail.

Comment (cbednarski) [inline client/allocdir/alloc_dir.go]: That makes sense. It might be easier to make linkOrCopy a bit smarter so it can no-op on a basic CRC or file mtime check or something.

File: client/task_runner_test.go [hunk 2]
```diff
@@ -172,28 +168,22 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
...
 	defer tr2.Destroy()
 
 	// Destroy and wait
-	tr2.Destroy()
-
-	select {
-	case <-tr.WaitCh():
-	case <-time.After(2 * time.Second):
-		t.Fatalf("timeout")
+	time.Sleep(200 * time.Millisecond)
```
Comment (cbednarski) [inline client/task_runner_test.go]: Because of the goroutine on 181 I think this is actually a race condition if we don't use a channel or wait group. Maybe:

```
go func() {
    tr2.Run()
    close(doneCh)
}()
```

File: client/alloc_runner.go [hunk 3]
```diff
@@ -103,11 +103,18 @@ func (r *AllocRunner) RestoreState() error {
...
-	for name := range r.taskStatus {
+	for name, status := range r.taskStatus {
 		task := &structs.Task{Name: name}
 		restartTracker := newRestartTracker(r.alloc.Job.Type, r.RestartPolicy)
 		tr := NewTaskRunner(r.logger, r.config, r.setTaskStatus, r.ctx, r.alloc.ID, task, restartTracker)
 		r.tasks[name] = tr
+
+		// Skip tasks in terminal states.
+		if status.Status == structs.AllocClientStatusDead ||
+			status.Status == structs.AllocClientStatusFailed {
```
Comment (diptanu) [inline client/alloc_runner.go]: Will AllocClientStatus ever get to failed state?

File: client/alloc_runner.go [hunk 4]
```diff
@@ -302,7 +309,7 @@ func (r *AllocRunner) Run() {
 	r.taskLock.Lock()
 	for _, task := range tg.Tasks {
 		// Skip tasks that were restored
-		if _, ok := r.tasks[task.Name]; ok {
+		if _, ok := r.taskStatus[task.Name]; ok {
```
Comment (diptanu) [inline client/alloc_runner.go]: @dadgar I think it would be clean to not start TaskRunners while restoring them. Restore state could just recreate the state of the structs, and when the client finishes calling the restoreState, it could call run on all the AllocRunner that it has, which would then call Run on all it's Task Runners? That way all the logic of how to run or attach to a Task is always encapsulated in the Run method. If the state of a Task is running the run method calls Open on a Driver otherwise it calls Start.

File: client/client.go [hunk 5]
```diff
@@ -148,8 +148,15 @@ func (c *Client) init() error {
 			return fmt.Errorf("failed creating state dir: %s", err)
 		}
 
-		c.logger.Printf("[INFO] client: using state directory %v", c.config.StateDir)
+	} else {
```
Comment (diptanu) [inline client/client.go]: If the user hasn't provided a state directory, perhaps he doesn't want checkpointing features of the Nomad client? Should we be even trying to persist state?

File: client/alloc_runner.go (hunk 3 above)
Comment (dadgar) [inline client/alloc_runner.go]: Yep you can see them here: https://github.com/hashicorp/nomad/blob/fbabd1237baf789cd0120d7ec13edb25813ae3a9/client/task_runner.go#L155

File: client/client.go (hunk 5 above)
Comment (dadgar) [inline client/client.go]: I think that is a separate question to be addressed in a different PR. Right now the system assumes you have a state_dir: https://github.com/hashicorp/nomad/blob/master/client/client.go#L319.

A different PR could disable state saving.

File: client/alloc_runner.go (hunk 3 above)
Comment (diptanu) [inline client/alloc_runner.go]: @dadgar What is the implication of this being failed? Ideally in this case if creating the driver failed, the server should mark this alloc as Dead and restart it elsewhere.

Comment (dadgar) [inline client/alloc_runner.go]: Yeah that is what happens. Failed just distinguishes from Dead that the user process never started.
